torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -0,0 +1,268 @@
+#include <queue>
+
+#include <torch/csrc/autograd/functions/accumulate_grad.h>
+#include <torch/csrc/autograd/input_buffer.h>
+#include <torch/csrc/distributed/autograd/context/dist_autograd_container.h>
+#include <torch/csrc/distributed/autograd/engine/dist_engine.h>
+
+namespace torch {
+namespace distributed {
+namespace autograd {
+
+using torch::autograd::AccumulateGrad;
+using torch::autograd::edge_list;
+using torch::autograd::Engine;
+using torch::autograd::GraphRoot;
+using torch::autograd::GraphTask;
+using torch::autograd::Node;
+using torch::autograd::validate_outputs;
+using torch::autograd::variable_list;
+
+DistEngine::DistEngine()
+    : initializedContextIds_(), engine_(Engine::get_default_engine()) {}
+
+DistEngine& DistEngine::getInstance() {
+  static DistEngine engine;
+  return engine;
+}
+
+void DistEngine::validateRootsAndRetrieveEdges(
+    const variable_list& roots,
+    edge_list& rootEdges,
+    variable_list& grads) {
+  TORCH_CHECK(!roots.empty(), "No tensors provided for gradient computation.");
+  TORCH_INTERNAL_ASSERT(rootEdges.empty());
+  TORCH_INTERNAL_ASSERT(grads.empty());
+
+  // Verify roots are all scalar and require gradients.
+  for (const auto& root : roots) {
+    TORCH_CHECK(
+        root.requires_grad(), "requires_grad not set on: ", root.name());
+    TORCH_CHECK(
+        root.numel() == 1,
+        root.name(),
+        " is not a scalar, all roots need to be scalar");
+    TORCH_CHECK(
+        root.grad_fn(),
+        root.name(),
+        " does not have a valid gradient function.");
+
+    // Compute the root edges and generate the appropriate gradients.
+    rootEdges.push_back(root.gradient_edge());
+    grads.push_back(at::ones_like(root));
+  }
+
+  // Validate rootEdges and grads.
+  validate_outputs(
+      rootEdges, grads, [](const std::string& msg) { return msg; });
+}
+
+void DistEngine::computeDependencies(
+    DistAutogradContext& autogradContext,
+    const edge_list& rootEdges,
+    const variable_list& grads,
+    std::shared_ptr<Node>& graphRoot,
+    edge_list& outputEdges) {
+  // Build the graph task and graph root.
+  auto graphTask = std::unique_ptr<GraphTask>(new GraphTask(
+      /* keep_graph */ false, /* create_graph */ false, /* depth */ 0));
+
+  if (graphRoot == nullptr) {
+    // Create a graphRoot only if needed.
```

Comment:
Hmm? (There's only two call sites for this function; surely we can describe in more detail the case when `graphRoot == nullptr`...)

Reply:
A better API (one that makes illegal states unrepresentable) is to just only take `graphRoot`, and force the client of the function to build the GraphRoot themselves. The illegal state you make unrepresentable in this case is when `graphRoot != null`, but `rootEdges` and `grads` don't agree with `graphRoot`.